### PR TITLE
Add dropout2d

### PR DESCRIPTION
**Pull request by deschena**
## Proposed changes

Implement 2D dropout + expose 1D dropout in documentation. I have tested my code against PyTorch's `Dropout2d`. Should I include further tests, e.g. comparing with torch results ?

## Checklist

Put an `x` in the boxes that apply.

- [x] I have read the [CONTRIBUTING](https://github.com/ml-explore/mlx/blob/main/CONTRIBUTING.md) document
- [x] I have run `pre-commit run --all-files` to format my code / installed pre-commit prior to committing changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] I have updated the necessary documentation (if needed)

### CODE REVIEW

File: docs/src/python/nn/layers.rst [hunk 1]
```diff
@@ -26,3 +26,5 @@ Layers
    RoPE
    MultiHeadAttention
    Sequential
+   Dropout
```
Comment (awni) [inline docs/src/python/nn/layers.rst:29]: 👍

File: python/mlx/nn/layers/dropout.py [hunk 2]
```diff
@@ -33,3 +33,45 @@ def __call__(self, x):
...
+    """Randomly zero an entire channel. During training, each channel
+    will be zeroed-out independently with probability p.
+
+    The remaining channels are multiplied with :math:`\frac{1}{1-p}` where
+    :math:`p` is the probability of zeroing an element. This is done so the
+    expected value of a given element will remain the same.
+
+    Args:
+        p (float): The probability to zero a channel
+    """
```
Comment (awni) [inline python/mlx/nn/layers/dropout.py:48]: - Could you say a bit more about the expected ordering / shape of the input?
- Also for our Conv2D we do `NHWC` by default, so probably we should change this to match that?

File: python/mlx/nn/layers/dropout.py [hunk 3]
```diff
@@ -33,3 +33,45 @@ def __call__(self, x):
...
+            return x
+
+        mask = mx.random.bernoulli(p=self._p_1, shape=x.shape[:2])
+        if x.ndim == 3:
+            mask = mask[..., None]
+
+        if x.ndim == 4:
+            mask = mask[..., None, None]
+
+        return (1 / self._p_1) * mask.astype(x.dtype) * x
```
Comment (awni) [inline python/mlx/nn/layers/dropout.py:77]: Why do a cast on the mask here? Shouldn't the `*` automatically upcast to the right type?

**Review by awni — CHANGES_REQUESTED** [review]
Nice! Thanks for adding this! Just a few comments / questions :)

File: python/mlx/nn/layers/dropout.py (hunk 2 above)
Comment (deschena) [reply python/mlx/nn/layers/dropout.py:48]: Yes, my bad, I missed that convolutions used a different convention 🥲 Will update

File: python/mlx/nn/layers/dropout.py (hunk 3 above)
Comment (deschena) [reply python/mlx/nn/layers/dropout.py:77]: I think it should, yes, I just saw it was done in the regular Dropout so I also did it.

Comment (awni) [reply python/mlx/nn/layers/dropout.py:85]: Cool, that's probably an oversight. Do you mind removing them both 😄 ?

Comment (deschena) [reply python/mlx/nn/layers/dropout.py]: Of course 👍

File: python/mlx/nn/layers/dropout.py [hunk 4]
```diff
@@ -32,4 +32,54 @@ def __call__(self, x):
...
         mask = mx.random.bernoulli(self._p_1, x.shape)
 
-        return (1 / self._p_1) * mask.astype(x.dtype) * x
+        return (1 / self._p_1) * mask * x
+
+
+class Dropout2d(Module):
+    """Apply 2D channel-wise dropout during training.
+
+    Randomly zero out entire channels independently with probability p.
```
Comment (awni) [inline python/mlx/nn/layers/dropout.py:41]: Can you put "p" in math mode for consistency with below:  :math:`p`

File: python/mlx/nn/layers/dropout.py [hunk 5]
```diff
@@ -32,4 +32,54 @@ def __call__(self, x):
...
+    This layer expects input of shape (B, W, H, C) or (W, H, C).
```
Comment (awni) [inline python/mlx/nn/layers/dropout.py:42]: Could you add a comment like: 
```
where `B` is the batch dimension, `C` is the channel dimension, and `H` and `W` are the height and width respectively
```

Also I would add double backticks around the shapes.

File: python/mlx/nn/layers/dropout.py [hunk 6]
```diff
@@ -32,4 +32,54 @@ def __call__(self, x):
...
+
+    Unlike traditional dropout, which zeros individual entries,
+    this layer zeros entire channels. This is beneficial for early
+    convolution layers where adjacent pixels are correlated. In such case,
+    traditional dropout may not effectively regularize activations. For more details, see [1]
```
Comment (awni) [inline python/mlx/nn/layers/dropout.py:47]: nit: period after [1]

**Review by awni — APPROVED** [review]
This looks great! Left a few more style nits for the docstring. Also can you rebase / resolve conflicts.

We can merge after those changes 🚀

**Comment by deschena** [comment]
Arg, I made a mess with the git history, I will deal with it asap

**Comment by awni** [comment]
I'm not sure on the state of your repo. But what usually works well for me is to:

1. Pull with rebase
2. Resolve conflicts during rebase
3. Commit and force push to the branch you are working

**Comment by deschena** [comment]
I dont know what I did, but now I cannot push to my branch anymore 🥲

**Comment by deschena** [comment]
Argh
